### ISSUE TITLE
visitedAny criteria: please check again

### ISSUE THREAD

**Issue by martindusinberre** [uzh/marugoto#11]
Lena are not sure the visitedAny criteria is working for pageTransition files. So far as Christian explained it (Rocketchat 11 July), if I have "visitedAny" followed by a, b, or c, the player will *only* see the pageTransition button if they were on a, b or c. But that isn't the way it's working in chapter4 at the moment: in the key page transitions (2.4.2.2, 2.4.3.2, 2.4.4.3), page transition buttons are visible even if the player was not on a, b, or c. This also seems to be a problem for "visited". Can you check again? (The functionality of notVisited is working just fine, so far as we can see.)

**Comment by martindusinberre** [comment]
An update on this: I have come into 2.4.2.2 from having visited HSA, ie having landed on 2.4.1.3.1.1HSA. I therefore should *not* be offered the option of going to the HSA. 
For 2.4.2.2/pageTransition4 (buttonText: Hawai'i State Archives), I therefore set the criteria as: "visitedAny" 
and 2.4.1.3.3.1Ham (ie, the HSA button text should be visible if the player visited 2.4.1.3.3.1Ham--the Hamilton Library)
and 2.4.1.3.4.1Ed (ie, the HSA button text should be visible if the player visited 2.4.1.3.4.1Ed--the Edersons)
In my understanding of visitedAny, this buttonText should *only* be visible to players *if* they have previously visited either the Hamilton Library or the Edersons. It should *not* be visible to players if they previously visited the HSA. 
So, either I've misunderstood visitedAny or there's an error here. Can you let me know? Thanks...

**Comment by martindusinberre** [comment]
Another update: I have come into 2.4.3.2 from (i) Hamilton and (ii) HPV. I should therefore *not* be offered either of these as options. And I also should *not* be offered HMCS, as I haven't yet been to HSA. I'm therefore not sure why I have been offered these three buttons.

**Comment by christian-bernet** [comment]
I tried it with page 2.4.2.2

**Comment by christian-bernet** [comment]
This is a bug. I have to fix it... :-)

**Comment by christian-bernet** [comment]
OK. I have found the bug. 
The content was not right. 
You have to use the attribute affectedPagesIds and not affectedPage for multiple pages for the visitedAny criteria.

I also have overseen that first.....
For page 2.4.2.2 it is changed now and it works.
Change pushed now to content.

**Comment by interrogator** [comment]
@martindusinberre Please let me know if you'd like validator to check for anything related to this.

**Comment by martindusinberre** [comment]
This is fixed in terms of visitedAny.
But there's now still a problem in terms of notVisited when there are more than one inputs.
For example, I played chapter 4 visiting the Edersons first (2.4.1.3.4.1Ed). 
When I got to 2.4.2.2, I was no longer offered the Edersons because the visitedAny functionality now works fine. Thanks!
But then, having played through HSA (2.4.3.1.1.1HSA) and come to 2.4.3.2, I was mistakenly offered the option of "Emi and Al Ederson". I had set this as follows:
notVisited 2.4.1.3.4.1Ed *and* notVisited 2.4.3.1.4.1Ed
In other words, I thought that players would _only_ be offered the Ederson option if they had _not_ visited the two previous Ederson pages.
Have I set this up wrongly? Perhaps I mean they should not be offered the Ederson option if they have notVisited *either* 2.4.1.3.4.1Ed *or* 2.4.3.1.4.1Ed.
I think we don't have the notVisitedAny functionality, so is there a workaround to this?

**Comment by martindusinberre** [comment]
The combination of notVisited is also a problem at 2.4.4.1.3.3, for example, when the player sees two "Continue" buttons (they should only see one).

**Comment by christian-bernet** [comment]
I have to check that.

**Comment by martindusinberre** [comment]
visitedAny isn't working again. The problem is on 2.4.3.2, and then on several pages after that. Basically, if I have come to 2.4.3.2 from either 2.4.1.3.3.1Ham or 2.4.3.1.3.1Ham, I should be offered the option of going to HPV (that is, 2.4.3.2/pageTransition3). But I am not being offered that option. 
Similarly, if I come from either 2.4.1.3.1.1HSA or 2.4.3.1.1.1HSA, I should be offered the option of going to HMCS (that is, 2.4.3.2/pageTransition1). But I'm not.
In both cases, this is visitedAny functionality.
Similar problem on the 2.4.4.3 transitions.

**Comment by martindusinberre** [comment]
I will do a work-around to work for Monday: I'll effectively cut the page criteria for what were until now visitedAny pageTransitions on 2.4.3.2 and 2.4.4.3 respectively. In other cases, e.g. 2.4.4.1.1.6 and similar variations (2.4.4.1.1.8, 2.4.4.1.2.4 etc) I've left the malfunctioning pageTransition file but simply added another one.

**Comment by christian-bernet** [comment]
Estimate 4 hours

**Comment by christian-bernet** [comment]
@martindusinberre can you please add now the wished page transition, so that I can have a look at it afterwards?
I did not find any transition on page  2.4.3.2 or 2.4.4.3 with criteria visitedAny.

**Comment by martindusinberre** [comment]
@christian-bernet I deleted all the visitedAny criteria before the test because they weren't working. I've now added some criteria to 2.4.3.2/pageTransition1.json and 2.4.3.2/pageTransition3.json. I've left the button text the same so that you can see the criteria I want: that is (on pT1) that the button text HMCS is **only** visible to the player **if** they visited **either** HSA on day 1 (2.4.1.3.1.1HSA) **or** on day 3 (2.4.3.1.1.1HSA).
Same for pT3: they should only have the button text for HPV if they visited Hamilton on either day 1 or day 3. 
Please check (I just pushed the changes now). If we can get visitedAny working on 2.4.3.2, then I can start implementing it elsewhere (2.4.4.3 will also need some rewriting)

**Comment by Helena-LiT** [comment]
visitedAny works for 2.4.3.2/pageTransition1.json and 2.4.3.2/pageTransition3.json. The old button text can go now, the the issue can be closed.
